# ç¼–ç è§„èŒƒ

## ğŸ“‹ Spring Boot 2.7 + Dubbo 3 ä¼ä¸šçº§ç¼–ç è§„èŒƒ

æœ¬æ–‡æ¡£å®šä¹‰äº† ai-coding-java é¡¹ç›®ä¸­çš„ç¼–ç æ ‡å‡†å’Œæœ€ä½³å®è·µï¼Œç¡®ä¿ä»£ç è´¨é‡ã€å¯ç»´æŠ¤æ€§å’Œå›¢é˜Ÿåä½œæ•ˆç‡ã€‚

## [TARGET] æ ¸å¿ƒåŸåˆ™

### 1. å¯è¯»æ€§ä¼˜å…ˆ
- ä»£ç åº”è¯¥æ˜“äºç†è§£å’Œç»´æŠ¤
- æ¸…æ™°çš„å‘½åå’Œæ³¨é‡Š
- é€»è¾‘ç®€å•ï¼Œé¿å…è¿‡åº¦ä¼˜åŒ–

### 2. ä¸€è‡´æ€§
- éµå¾ªå›¢é˜Ÿç»Ÿä¸€çš„ç¼–ç é£æ ¼
- ä½¿ç”¨ç›¸åŒçš„è®¾è®¡æ¨¡å¼
- ä¿æŒä»£ç ç»“æ„çš„ä¸€è‡´æ€§

### 3. å¯æµ‹è¯•æ€§
- ç¼–å†™å¯å•å…ƒæµ‹è¯•çš„ä»£ç 
- ä¾èµ–æ³¨å…¥å’Œé¢å‘æ¥å£ç¼–ç¨‹
- é¿å…ç¡¬ç¼–ç å’Œé™æ€ä¾èµ–

## [EDIT] å‘½åè§„èŒƒ

### 1. åŒ…å‘½å
```java
// [OK] æ­£ç¡® - å…¨å°å†™ï¼Œç”¨ç‚¹å·åˆ†éš”
package com.company.project.module.service.impl;

// [X] é”™è¯¯ - é¦–å­—æ¯å¤§å†™æˆ–ä½¿ç”¨ä¸‹åˆ’çº¿
package com.Company.Project.Module.Service;
package com.company.project.module.service_impl;
```

### 2. ç±»å‘½å
```java
// [OK] æ­£ç¡® - å¤§é©¼å³°å‘½åæ³•
public class UserService {}
public class OrderRepositoryImpl {}
public class PaymentController {}

// [X] é”™è¯¯ - å°é©¼å³°æˆ–ä¸‹åˆ’çº¿
public class userService {}
public class order_repository_impl {}
```

### 3. æ–¹æ³•å’Œå˜é‡å‘½å
```java
public class UserService {

    // [OK] æ­£ç¡® - å°é©¼å³°å‘½åæ³•
    private UserRepository userRepository;
    private String userName;
    private List<Order> orderList;

    // [OK] æ­£ç¡® - åŠ¨è¯å¼€å¤´ï¼Œè¡¨è¾¾æ˜ç¡®çš„æ“ä½œ
    public User createUser(CreateUserRequest request) {}
    public boolean validateUserCredentials(String username, String password) {}
    public List<User> findActiveUsersByDepartment(String departmentId) {}

    // [X] é”™è¯¯ - åè¯å¼€å¤´æˆ–å«ä¹‰ä¸æ˜
    public User user(CreateUserRequest request) {}
    public boolean check(String username, String password) {}
    public List<User> get(String departmentId) {}
}
```

### 4. å¸¸é‡å‘½å
```java
public class OrderConstants {

    // [OK] æ­£ç¡® - å…¨å¤§å†™ï¼Œä¸‹åˆ’çº¿åˆ†éš”
    public static final int MAX_ORDER_AMOUNT = 10000;
    public static final String DEFAULT_STATUS = "PENDING";
    public static final long TOKEN_EXPIRE_TIME = 3600L;

    // [X] é”™è¯¯ - å°é©¼å³°æˆ–æ··åˆå¤§å°å†™
    public static final int maxOrderAmount = 10000;
    public static final String defaultStatus = "PENDING";
}
```

## [BOOK] æ³¨é‡Šè§„èŒƒ

### 1. ç±»æ³¨é‡Š
```java
/**
 * ç”¨æˆ·æœåŠ¡å®ç°ç±»
 *
 * æä¾›ç”¨æˆ·ç®¡ç†ç›¸å…³çš„ä¸šåŠ¡é€»è¾‘å®ç°ï¼ŒåŒ…æ‹¬ç”¨æˆ·æ³¨å†Œã€ç™»å½•ã€ä¿¡æ¯æ›´æ–°ç­‰åŠŸèƒ½ã€‚
 * æ”¯æŒåˆ†å¸ƒå¼éƒ¨ç½²ï¼Œé€šè¿‡ Dubbo æä¾›æœåŠ¡æ¥å£ã€‚
 *
 * @author å¼€å‘è€…å§“å
 * @version 1.0.0
 * @since 2023-12-07
 */
@Service
@Transactional
public class UserServiceImpl implements UserService {

}
```

### 2. æ–¹æ³•æ³¨é‡Š
```java
/**
 * åˆ›å»ºæ–°ç”¨æˆ·
 *
 * æ ¹æ®æä¾›çš„ç”¨æˆ·ä¿¡æ¯åˆ›å»ºæ–°ç”¨æˆ·è´¦æˆ·ï¼ŒåŒ…æ‹¬ï¼š
 * - åŸºæœ¬ä¿¡æ¯éªŒè¯å’Œå­˜å‚¨
 * - å¯†ç åŠ å¯†
 * - é»˜è®¤è§’è‰²åˆ†é…
 * - å®¡è®¡ä¿¡æ¯è®°å½•
 *
 * @param request ç”¨æˆ·åˆ›å»ºè¯·æ±‚ï¼ŒåŒ…å«ç”¨æˆ·åŸºæœ¬ä¿¡æ¯
 * @return åˆ›å»ºæˆåŠŸçš„ç”¨æˆ·ä¿¡æ¯ï¼Œä¸åŒ…å«æ•æ„Ÿæ•°æ®
 * @throws BusinessException å½“ç”¨æˆ·åå·²å­˜åœ¨æ—¶æŠ›å‡º
 * @throws ValidationException å½“è¯·æ±‚æ•°æ®éªŒè¯å¤±è´¥æ—¶æŠ›å‡º
 */
@Transactional(rollbackFor = Exception.class)
public UserResponse createUser(CreateUserRequest request) {
    // å®ç°é€»è¾‘
}
```

### 3. å¤æ‚é€»è¾‘æ³¨é‡Š
```java
public Order calculateOrderTotal(Order order) {
    BigDecimal totalAmount = BigDecimal.ZERO;

    // è®¡ç®—å•†å“æ€»ä»·
    for (OrderItem item : order.getItems()) {
        // å•†å“å•ä»· Ã— æ•°é‡ - ä¿ƒé”€æŠ˜æ‰£
        BigDecimal itemTotal = item.getUnitPrice()
            .multiply(new BigDecimal(item.getQuantity()))
            .subtract(item.getDiscountAmount());
        totalAmount = totalAmount.add(itemTotal);
    }

    // åº”ç”¨è®¢å•çº§ä¼˜æƒ åˆ¸æŠ˜æ‰£
    if (order.getCouponCode() != null) {
        Coupon coupon = couponService.getByCode(order.getCouponCode());
        if (coupon != null && coupon.isValid()) {
            totalAmount = totalAmount.subtract(coupon.getDiscountAmount());
        }
    }

    // æ£€æŸ¥æœ€å°è®¢å•é‡‘é¢é™åˆ¶
    if (totalAmount.compareTo(BigDecimal.valueOf(10)) < 0) {
        throw new BusinessException("è®¢å•é‡‘é¢ä¸èƒ½å°äº10å…ƒ");
    }

    return totalAmount;
}
```

## [ARCHITECTURE] ä»£ç ç»“æ„è§„èŒƒ

### 1. ç±»ç»“æ„é¡ºåº
```java
public class UserService {

    // 1. é™æ€å¸¸é‡
    private static final Logger logger = LoggerFactory.getLogger(UserService.class);

    // 2. å®ä¾‹å˜é‡ï¼ˆæŒ‰è®¿é—®çº§åˆ«æ’åºï¼šprivate -> protected -> publicï¼‰
    @Autowired
    private UserRepository userRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    // 3. æ„é€ æ–¹æ³•
    public UserService() {}

    @Autowired
    public UserService(UserRepository userRepository, PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
    }

    // 4. å…¬å…±æ–¹æ³•ï¼ˆæŒ‰ä¸šåŠ¡é€»è¾‘åˆ†ç»„ï¼‰
    public User createUser(CreateUserRequest request) {}
    public User updateUser(Long userId, UpdateUserRequest request) {}
    public void deleteUser(Long userId) {}

    // 5. å—ä¿æŠ¤æ–¹æ³•
    protected void validateUserRequest(CreateUserRequest request) {}

    // 6. ç§æœ‰æ–¹æ³•
    private String encryptPassword(String rawPassword) {}
    private UserResponse convertToResponse(User user) {}

    // 7. é™æ€æ–¹æ³•
    public static boolean isValidEmail(String email) {}
}
```

### 2. æ–¹æ³•é•¿åº¦æ§åˆ¶
```java
// [OK] æ­£ç¡® - å•ä¸€èŒè´£ï¼Œé•¿åº¦é€‚ä¸­
public UserResponse createUser(CreateUserRequest request) {
    // éªŒè¯è¯·æ±‚å‚æ•°
    validateCreateUserRequest(request);

    // æ£€æŸ¥ç”¨æˆ·åæ˜¯å¦å·²å­˜åœ¨
    if (userRepository.existsByUsername(request.getUsername())) {
        throw new BusinessException("ç”¨æˆ·åå·²å­˜åœ¨");
    }

    // åˆ›å»ºç”¨æˆ·å®ä½“
    User user = buildUserFromRequest(request);

    // ä¿å­˜ç”¨æˆ·
    User savedUser = userRepository.save(user);

    // å‘é€æ¬¢è¿é‚®ä»¶ï¼ˆå¼‚æ­¥ï¼‰
    sendWelcomeEmail(savedUser);

    return convertToResponse(savedUser);
}

// [X] é”™è¯¯ - æ–¹æ³•è¿‡é•¿ï¼ŒèŒè´£ä¸æ¸…
public UserResponse createUser(CreateUserRequest request) {
    // 100+ è¡Œä»£ç åŒ…å«éªŒè¯ã€ä¸šåŠ¡é€»è¾‘ã€æ•°æ®è½¬æ¢ã€é€šçŸ¥ç­‰
}
```

## ğŸ” å¼‚å¸¸å¤„ç†è§„èŒƒ

### 1. è‡ªå®šä¹‰å¼‚å¸¸ç±»
```java
// ä¸šåŠ¡å¼‚å¸¸åŸºç±»
public class BusinessException extends RuntimeException {
    private final String errorCode;

    public BusinessException(String message) {
        super(message);
        this.errorCode = "BUSINESS_ERROR";
    }

    public BusinessException(String errorCode, String message) {
        super(message);
        this.errorCode = errorCode;
    }
}

// å…·ä½“ä¸šåŠ¡å¼‚å¸¸
public class UserNotFoundException extends BusinessException {
    public UserNotFoundException(String username) {
        super("USER_NOT_FOUND", "ç”¨æˆ·ä¸å­˜åœ¨: " + username);
    }
}
```

### 2. å¼‚å¸¸å¤„ç†ç­–ç•¥
```java
@Service
public class UserService {

    public User createUser(CreateUserRequest request) {
        try {
            // ä¸šåŠ¡é€»è¾‘
            return userRepository.save(user);
        } catch (DataIntegrityViolationException e) {
            // æ•°æ®å®Œæ•´æ€§çº¦æŸå¼‚å¸¸
            throw new BusinessException("USER_ALREADY_EXISTS", "ç”¨æˆ·å·²å­˜åœ¨");
        } catch (Exception e) {
            // å…¶ä»–æœªçŸ¥å¼‚å¸¸
            logger.error("åˆ›å»ºç”¨æˆ·å¤±è´¥", e);
            throw new BusinessException("SYSTEM_ERROR", "ç³»ç»Ÿé”™è¯¯ï¼Œè¯·ç¨åé‡è¯•");
        }
    }
}
```

### 3. å…¨å±€å¼‚å¸¸å¤„ç†
```java
@RestControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(BusinessException.class)
    public ResponseEntity<ErrorResponse> handleBusinessException(BusinessException e) {
        ErrorResponse error = ErrorResponse.builder()
            .code(e.getErrorCode())
            .message(e.getMessage())
            .timestamp(Instant.now())
            .build();
        return ResponseEntity.badRequest().body(error);
    }

    @ExceptionHandler(ValidationException.class)
    public ResponseEntity<ErrorResponse> handleValidationException(ValidationException e) {
        ErrorResponse error = ErrorResponse.builder()
            .code("VALIDATION_ERROR")
            .message(e.getMessage())
            .timestamp(Instant.now())
            .build();
        return ResponseEntity.badRequest().body(error);
    }
}
```

## ğŸ§ª æµ‹è¯•è§„èŒƒ

### 1. å•å…ƒæµ‹è¯•
```java
@ExtendWith(MockitoExtension.class)
class UserServiceTest {

    @Mock
    private UserRepository userRepository;

    @Mock
    private PasswordEncoder passwordEncoder;

    @InjectMocks
    private UserService userService;

    @Test
    @DisplayName("åˆ›å»ºç”¨æˆ·æˆåŠŸ")
    void shouldCreateUserSuccessfully() {
        // Given
        CreateUserRequest request = CreateUserRequest.builder()
            .username("testuser")
            .email("test@example.com")
            .password("password123")
            .build();

        User savedUser = User.builder()
            .id(1L)
            .username("testuser")
            .email("test@example.com")
            .password("encryptedPassword")
            .build();

        when(userRepository.existsByUsername("testuser")).thenReturn(false);
        when(passwordEncoder.encode("password123")).thenReturn("encryptedPassword");
        when(userRepository.save(any(User.class))).thenReturn(savedUser);

        // When
        UserResponse response = userService.createUser(request);

        // Then
        assertThat(response.getUsername()).isEqualTo("testuser");
        assertThat(response.getEmail()).isEqualTo("test@example.com");
        verify(userRepository).save(any(User.class));
    }

    @Test
    @DisplayName("ç”¨æˆ·åå·²å­˜åœ¨æ—¶æŠ›å‡ºå¼‚å¸¸")
    void shouldThrowExceptionWhenUsernameExists() {
        // Given
        CreateUserRequest request = CreateUserRequest.builder()
            .username("existinguser")
            .build();

        when(userRepository.existsByUsername("existinguser")).thenReturn(true);

        // When & Then
        assertThatThrownBy(() -> userService.createUser(request))
            .isInstanceOf(BusinessException.class)
            .hasMessage("ç”¨æˆ·åå·²å­˜åœ¨");
    }
}
```

### 2. é›†æˆæµ‹è¯•
```java
@SpringBootTest
@TestPropertySource(properties = {
    "spring.datasource.url=jdbc:h2:mem:testdb",
    "spring.jpa.hibernate.ddl-auto=create-drop"
})
class UserControllerIntegrationTest {

    @Autowired
    private TestRestTemplate restTemplate;

    @Test
    void shouldCreateUserSuccessfully() {
        // Given
        CreateUserRequest request = CreateUserRequest.builder()
            .username("testuser")
            .email("test@example.com")
            .password("password123")
            .build();

        // When
        ResponseEntity<UserResponse> response = restTemplate.postForEntity(
            "/api/users", request, UserResponse.class);

        // Then
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.CREATED);
        assertThat(response.getBody().getUsername()).isEqualTo("testuser");
    }
}
```

## [TOOL] é…ç½®è§„èŒƒ

### 1. é…ç½®ç±»
```java
@Configuration
@EnableConfigurationProperties({RedisProperties.class, DatabaseProperties.class})
public class AppConfig {

    @Bean
    @ConditionalOnProperty(name = "cache.enabled", havingValue = "true")
    public CacheManager cacheManager(RedisConnectionFactory connectionFactory) {
        RedisCacheConfiguration config = RedisCacheConfiguration.defaultCacheConfig()
            .entryTtl(Duration.ofMinutes(30))
            .serializeKeysWith(RedisSerializationContext.SerializationPair.fromSerializer(new StringRedisSerializer()))
            .serializeValuesWith(RedisSerializationContext.SerializationPair.fromSerializer(new GenericJackson2JsonRedisSerializer()));

        return RedisCacheManager.builder(connectionFactory)
            .cacheDefaults(config)
            .build();
    }
}
```

### 2. é…ç½®å±æ€§
```java
@ConfigurationProperties(prefix = "app")
@Data
public class AppConfigProperties {

    private String name;
    private String version;
    private Cache cache = new Cache();

    @Data
    public static class Cache {
        private boolean enabled = true;
        private Duration ttl = Duration.ofMinutes(30);
        private int maxSize = 1000;
    }
}
```

## [CHART] æ€§èƒ½ä¼˜åŒ–è§„èŒƒ

### 1. æ•°æ®åº“è®¿é—®ä¼˜åŒ–
```java
// [OK] æ­£ç¡® - ä½¿ç”¨åˆ†é¡µæŸ¥è¯¢
public Page<User> findUsersByPage(UserQuery query, Pageable pageable) {
    return userRepository.findByConditions(
        query.getUsername(),
        query.getEmail(),
        query.getStatus(),
        pageable
    );
}

// [X] é”™è¯¯ - ä¸€æ¬¡æ€§åŠ è½½å¤§é‡æ•°æ®
public List<User> findAllUsers() {
    return userRepository.findAll(); // å¯èƒ½å¯¼è‡´å†…å­˜æº¢å‡º
}
```

### 2. ç¼“å­˜ä½¿ç”¨è§„èŒƒ
```java
@Service
public class UserService {

    @Cacheable(value = "users", key = "#userId")
    public User getUserById(Long userId) {
        return userRepository.findById(userId)
            .orElseThrow(() -> new UserNotFoundException(userId));
    }

    @CacheEvict(value = "users", key = "#user.id")
    public User updateUser(User user) {
        return userRepository.save(user);
    }
}
```

### 3. å¼‚æ­¥å¤„ç†
```java
@Service
public class NotificationService {

    @Async
    public void sendEmailAsync(String to, String subject, String content) {
        try {
            emailService.send(to, subject, content);
        } catch (Exception e) {
            logger.error("å‘é€é‚®ä»¶å¤±è´¥: to={}, subject={}", to, subject, e);
        }
    }
}
```

## ğŸ”’ å®‰å…¨ç¼–ç è§„èŒƒ

### 1. è¾“å…¥éªŒè¯
```java
public class CreateUserRequest {

    @NotBlank(message = "ç”¨æˆ·åä¸èƒ½ä¸ºç©º")
    @Size(min = 3, max = 50, message = "ç”¨æˆ·åé•¿åº¦å¿…é¡»åœ¨3-50ä¸ªå­—ç¬¦ä¹‹é—´")
    @Pattern(regexp = "^[a-zA-Z0-9_]+$", message = "ç”¨æˆ·ååªèƒ½åŒ…å«å­—æ¯ã€æ•°å­—å’Œä¸‹åˆ’çº¿")
    private String username;

    @NotBlank(message = "é‚®ç®±ä¸èƒ½ä¸ºç©º")
    @Email(message = "é‚®ç®±æ ¼å¼ä¸æ­£ç¡®")
    private String email;

    @NotBlank(message = "å¯†ç ä¸èƒ½ä¸ºç©º")
    @Size(min = 8, message = "å¯†ç é•¿åº¦ä¸èƒ½å°‘äº8ä½")
    @Pattern(regexp = "^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[@$!%*?&])[A-Za-z\\d@$!%*?&]",
             message = "å¯†ç å¿…é¡»åŒ…å«å¤§å°å†™å­—æ¯ã€æ•°å­—å’Œç‰¹æ®Šå­—ç¬¦")
    private String password;
}
```

### 2. æ•æ„Ÿä¿¡æ¯å¤„ç†
```java
public class UserResponse {

    private Long id;
    private String username;
    private String email;

    // [X] ä¸è¦åœ¨å“åº”ä¸­åŒ…å«æ•æ„Ÿä¿¡æ¯
    // private String password;
    // private String creditCardNumber;

    // [OK] ä½¿ç”¨æ•°æ®è„±æ•
    public String getPhone() {
        return phone != null ? phone.replaceAll("(\\d{3})\\d{4}(\\d{4})", "$1****$2") : null;
    }
}
```

### 3. SQL æ³¨å…¥é˜²æŠ¤
```java
// [OK] æ­£ç¡® - ä½¿ç”¨å‚æ•°åŒ–æŸ¥è¯¢
@Query("SELECT u FROM User u WHERE u.username = :username AND u.status = :status")
List<User> findByUsernameAndStatus(@Param("username") String username, @Param("status") String status);

// [X] é”™è¯¯ - å­—ç¬¦ä¸²æ‹¼æ¥å®¹æ˜“å¯¼è‡´ SQL æ³¨å…¥
// @Query("SELECT u FROM User u WHERE u.username = '" + username + "'")
```

## ğŸ“‹ ä»£ç å®¡æŸ¥æ£€æŸ¥æ¸…å•

### 1. å‘½åå’Œå¯è¯»æ€§
- [ ] ç±»åã€æ–¹æ³•åã€å˜é‡åç¬¦åˆå‘½åè§„èŒƒ
- [ ] å¸¸é‡ä½¿ç”¨å…¨å¤§å†™ä¸‹åˆ’çº¿åˆ†éš”
- [ ] ä»£ç è‡ªæ–‡æ¡£åŒ–ï¼Œå¿…è¦æ—¶æœ‰æ³¨é‡Š
- [ ] å¤æ‚é€»è¾‘æœ‰è¯¦ç»†æ³¨é‡Šè¯´æ˜

### 2. å¼‚å¸¸å¤„ç†
- [ ] ä¸šåŠ¡å¼‚å¸¸ä½¿ç”¨è‡ªå®šä¹‰å¼‚å¸¸ç±»
- [ ] å¼‚å¸¸ä¿¡æ¯æ¸…æ™°æ˜ç¡®
- [ ] æœ‰å…¨å±€å¼‚å¸¸å¤„ç†æœºåˆ¶
- [ ] æ—¥å¿—è®°å½•æ°å½“ï¼Œä¸æ³„éœ²æ•æ„Ÿä¿¡æ¯

### 3. å®‰å…¨æ€§
- [ ] è¾“å…¥å‚æ•°éªŒè¯å®Œæ•´
- [ ] æ•æ„Ÿä¿¡æ¯ä¸è®°å½•åˆ°æ—¥å¿—
- [ ] SQL æ³¨å…¥é˜²æŠ¤åˆ°ä½
- [ ] æƒé™æ§åˆ¶æ­£ç¡®å®ç°

### 4. æ€§èƒ½è€ƒè™‘
- [ ] é¿å…ä¸å¿…è¦çš„æ•°æ®åº“æŸ¥è¯¢
- [ ] åˆç†ä½¿ç”¨ç¼“å­˜
- [ ] å¤§æ•°æ®é‡æ“ä½œä½¿ç”¨åˆ†é¡µ
- [ ] åŠæ—¶é‡Šæ”¾èµ„æº

### 5. æµ‹è¯•è¦†ç›–
- [ ] å•å…ƒæµ‹è¯•è¦†ç›–æ ¸å¿ƒä¸šåŠ¡é€»è¾‘
- [ ] è¾¹ç•Œæ¡ä»¶æµ‹è¯•å®Œæ•´
- [ ] å¼‚å¸¸åœºæ™¯æµ‹è¯•è¦†ç›–
- [ ] é›†æˆæµ‹è¯•éªŒè¯ç«¯åˆ°ç«¯æµç¨‹

---

## [OK] éµå¾ªæœ¬è§„èŒƒ

éµå¾ªæœ¬ç¼–ç è§„èŒƒå°†æœ‰åŠ©äºï¼š

1. **æé«˜ä»£ç è´¨é‡** - ç»Ÿä¸€çš„ç¼–ç æ ‡å‡†å‡å°‘é”™è¯¯
2. **å¢å¼ºå¯ç»´æŠ¤æ€§** - æ¸…æ™°çš„å‘½åå’Œç»“æ„ä¾¿äºç†è§£å’Œä¿®æ”¹
3. **ä¿ƒè¿›å›¢é˜Ÿåä½œ** - ç»Ÿä¸€çš„é£æ ¼å‡å°‘æ²Ÿé€šæˆæœ¬
4. **ä¿è¯ç³»ç»Ÿç¨³å®šæ€§** - è§„èŒƒçš„å¼‚å¸¸å¤„ç†å’Œå®‰å…¨é˜²æŠ¤
5. **æå‡å¼€å‘æ•ˆç‡** - æ ‡å‡†åŒ–çš„å¼€å‘å’Œæµ‹è¯•æµç¨‹

è¯·æ‰€æœ‰å¼€å‘äººå‘˜ä¸¥æ ¼éµå¾ªæœ¬è§„èŒƒï¼Œå¹¶åœ¨ä»£ç å®¡æŸ¥ä¸­ä½œä¸ºé‡è¦å‚è€ƒæ ‡å‡†ã€‚